core/utils/utils.go

```diff
@@ -866,65 +866,47 @@ const (
 
 // StartOnce sets the state to Started
 func (once *StartStopOnce) StartOnce(name string, fn func() error) error {
-	once.Lock()
-	defer once.Unlock()
+	success := once.state.CAS(int32(StartStopOnce_Unstarted), int32(StartStopOnce_Started))
 
-	if once.state != StartStopOnce_Unstarted {
+	if !success {
 		return errors.Errorf("%v has already started once", name)
 	}
-	once.state = StartStopOnce_Started
 
 	return fn()
```

Comment:
Hmm, might not be ideal if stop is called while the function is still executing. I could keep a `Mutex` specifically for `Start`/`Stop` calls (locking after the `CAS`), that way at least `IfStarted`/`Status` doesn't need to hold a lock.

Reply:
I've done that, though I'm not sure if it's really necessary in practice.